<a href="https://colab.research.google.com/github/syahrinalfi/MachineLearningTasks/blob/main/10thWeekTasks/10_DL_Pytorch_Boston_Housing.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

* Nama : Muhammad Alfi Syahrin
* NIM : 1103213144
* Lecture : 10
* Model : Deep Learning
* Dataset : [Boston Housing Dataset](https://www.openml.org/search?type=data&sort=runs&id=531&status=active)


# Import Library

In [1]:
import torch
import torch.nn as nn
import torch.optim as optim
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.datasets import fetch_openml

# Load Dataset

In [2]:
boston = fetch_openml(data_id=531)

/usr/local/lib/python3.10/dist-packages/sklearn/datasets/_openml.py:968: FutureWarning: The default value of `parser` will change from `'liac-arff'` to `'auto'` in 1.4. You can set `parser='auto'` to silence this warning. Therefore, an `ImportError` will be raised from 1.4 if the dataset is dense and pandas is not installed. Note that the pandas parser may return different data types. See the Notes Section in fetch_openml's API doc for details.
  warn(


In [3]:
# Load data dan target
X, y = boston.data, boston.target

# Preproccess Data

In [4]:
# Split dataset menjadi train dan test
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [5]:
# Standarisasi fitur
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

In [6]:
# Mengubah Data menjadi PyTorch tensors
X_train_tensor = torch.tensor(X_train, dtype=torch.float32)
y_train_tensor = torch.tensor(y_train.values, dtype=torch.float32).view(-1, 1)
X_test_tensor = torch.tensor(X_test, dtype=torch.float32)
y_test_tensor = torch.tensor(y_test, dtype=torch.float32).view(-1, 1)

# Model Train

In [7]:
# Membuat arsitektur neural network yang digunakan
class BostonModel(nn.Module):
    def __init__(self, input_size):
        super(BostonModel, self).__init__()
        self.fc1 = nn.Linear(input_size, 64)
        self.fc2 = nn.Linear(64, 32)
        self.fc3 = nn.Linear(32, 1)

    def forward(self, x):
        x = torch.relu(self.fc1(x))
        x = torch.relu(self.fc2(x))
        x = self.fc3(x)
        return x

In [8]:
# Mencontohkan Model
model = BostonModel(X_train.shape[1])

In [9]:
# Define loss dan optimizer
criterion = nn.MSELoss()
optimizer = optim.Adam(model.parameters(), lr=0.001)

In [10]:
# Train model
num_epochs = 100
for epoch in range(num_epochs):
    model.train()
    optimizer.zero_grad()
    outputs = model(X_train_tensor)
    loss = criterion(outputs, y_train_tensor)
    loss.backward()
    optimizer.step()
    if (epoch+1) % 10 == 0:
        print(f'Epoch [{epoch+1}/{num_epochs}], Loss: {loss.item():.4f}')

Epoch [10/100], Loss: 595.2652
Epoch [20/100], Loss: 577.1984
Epoch [30/100], Loss: 552.5729
Epoch [40/100], Loss: 518.2521
Epoch [50/100], Loss: 471.0692
Epoch [60/100], Loss: 408.9100
Epoch [70/100], Loss: 332.5201
Epoch [80/100], Loss: 247.7279
Epoch [90/100], Loss: 167.9282
Epoch [100/100], Loss: 109.3278


Dapat dilihat bahwa model yang sudah dilatih memiliki performa dengan nilai loss MSE sebesar 109.3278

In [11]:
# Evaluasi model
model.eval()
with torch.no_grad():
    y_pred = model(X_test_tensor)
    test_loss = criterion(y_pred, y_test_tensor)
    print(f'Test Loss: {test_loss.item():.4f}')

Test Loss: 84.2715


Dapat dilihat bahwa model melakukan prediksi terhadap data test dengan nilai loss MSE sebesar 84.2715